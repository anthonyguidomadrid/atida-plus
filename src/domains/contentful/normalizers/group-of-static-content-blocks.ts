import type { Entry } from 'contentful'
import { CmsContentTypes } from '~config/content-types'
import { removeUndefinedPropertiesFromObject } from '~helpers'
import {
  GroupOfStaticContentBlocksFragment,
  StaticContentBlockFragment,
  ContentBlockWithImage as ContentBlockWithImageFragment
} from '~generated-graphql'
import {
  StaticContentBlock,
  normalizeStaticContentBlockGraphQL
} from './static-content-block'
import {
  ContentBlockWithImage,
  normalizeContentBlockWithImageGraphQL
} from './content-block-with-image'

export type GroupOfStaticContentBlocks = {
  contentType: CmsContentTypes.GROUP_OF_STATIC_CONTENT_BLOCKS
  title?: string
  blocks?: (StaticContentBlock | ContentBlockWithImage)[]
  type: string
}
export type GroupOfStaticContentBlocksTypes =
  | 'Bullet points'
  | 'Accordion'
  | 'Slider'
export type ContentfulGroupOfStaticContentBlocks = Entry<{
  title?: string
  contentBlocks?: (StaticContentBlockFragment | ContentBlockWithImageFragment)[]
  type: GroupOfStaticContentBlocksTypes
}>

const normalizeGroupOfStaticContentBlocksGraphQL = (
  contentBlock: GroupOfStaticContentBlocksFragment
): GroupOfStaticContentBlocks | undefined => {
  return contentBlock
    ? removeUndefinedPropertiesFromObject({
        contentType: CmsContentTypes.GROUP_OF_STATIC_CONTENT_BLOCKS as const,
        title: contentBlock?.title ?? undefined,
        blocks:
          contentBlock?.staticContentBlocksCollection?.items &&
          contentBlock?.staticContentBlocksCollection?.items
            .map(block => {
              if (!block) {
                return null
              }
              if (
                (contentBlock.type === 'Accordion' ||
                  contentBlock.type === 'Bullet points') &&
                block.__typename === 'StaticContentBlock'
              ) {
                return normalizeStaticContentBlockGraphQL(block) ?? null
              }
              if (
                contentBlock.type === 'Slider' &&
                block.__typename === 'ContentBlockWithImage'
              ) {
                return normalizeContentBlockWithImageGraphQL(block) ?? null
              }
            })
            .filter(
              (block): block is StaticContentBlock | ContentBlockWithImage =>
                !!block
            ),
        type: contentBlock.type ?? ''
      })
    : undefined
}

export const normalizeGroupOfStaticContentBlocks = (
  groupOfStaticContentBlocks: GroupOfStaticContentBlocksFragment
): GroupOfStaticContentBlocks | undefined => {
  return normalizeGroupOfStaticContentBlocksGraphQL(groupOfStaticContentBlocks)
}
