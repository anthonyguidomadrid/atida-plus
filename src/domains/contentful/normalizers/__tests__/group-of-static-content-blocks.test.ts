import { normalizeGroupOfStaticContentBlocks } from '../group-of-static-content-blocks'

const contentBlockWithImage = {
  __typename: 'ContentBlockWithImage',
  title: 'Any questions?',
  image: {
    __typename: 'Asset',
    title: 'Contact image',
    description: 'Person looking at a tablet',
    url: 'https://atida.com/adam-winger-iirbrh939yc-unsplash_1.jpg',
    contentType: 'image/jpeg'
  },
  content: {
    __typename: 'ContentBlockWithImageContent',
    json: {
      data: {},
      content: [
        {
          data: {},
          content: [
            {
              data: {},
              marks: [],
              value:
                "We're here to help. Ask about our products or get expert medical advice.",
              nodeType: 'text'
            }
          ],
          nodeType: 'paragraph'
        }
      ],
      nodeType: 'document'
    }
  },
  cta: {
    __typename: 'Link',
    label: 'Chat with an expert',
    url: '/contact',
    icon: {
      __typename: 'Icon',
      ref: 'NavAdvice24'
    },
    content: null
  },
  textLink: {
    __typename: 'Link',
    label: 'Read our FAQ',
    url: '/faq',
    icon: null,
    content: null
  },
  bgColor: {
    __typename: 'Color',
    ref: 'secondary-dark-sky-blue'
  },
  textColor: {
    __typename: 'Color',
    ref: 'primary-white'
  },
  imageOnTheLeft: false,
  isSponsoredContent: false
} as const

const graphQLMockAccordion = {
  __typename: 'GroupOfStaticContentBlocks',
  title: 'Frequently asked questionsrs',
  type: 'Accordion',
  staticContentBlocksCollection: {
    items: [
      {
        __typename: 'StaticContentBlock',
        title: 'How do I earn Atida cash rewards?',
        content: {
          json: {
            nodeType: 'document',
            data: {},
            content: [
              {
                nodeType: 'heading-2',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value: 'Nam venenatis',
                    marks: [],
                    data: {}
                  }
                ]
              },
              {
                nodeType: 'paragraph',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value:
                      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.',
                    marks: [],
                    data: {}
                  }
                ]
              }
            ]
          }
        }
      },
      {
        __typename: 'StaticContentBlock',
        title: 'What are Atida cash rewards??',
        content: {
          json: {
            nodeType: 'document',
            data: {},
            content: [
              {
                nodeType: 'heading-2',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value: 'Nam venenatis',
                    marks: [],
                    data: {}
                  }
                ]
              },
              {
                nodeType: 'paragraph',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value:
                      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.',
                    marks: [],
                    data: {}
                  }
                ]
              }
            ]
          }
        }
      }
    ]
  }
} as const

const graphQLMockBulletPoints = {
  __typename: 'GroupOfStaticContentBlocks',
  title: 'Frequently asked questionsrs',
  type: 'Bullet points',
  staticContentBlocksCollection: {
    items: [
      {
        __typename: 'StaticContentBlock',
        title: 'How do I earn Atida cash rewards?',
        content: {
          json: {
            nodeType: 'document',
            data: {},
            content: [
              {
                nodeType: 'heading-2',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value: 'Nam venenatis',
                    marks: [],
                    data: {}
                  }
                ]
              },
              {
                nodeType: 'paragraph',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value:
                      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.',
                    marks: [],
                    data: {}
                  }
                ]
              }
            ]
          }
        },
        icon: {
          __typename: 'Icon',
          ref: 'iconReference'
        }
      },
      {
        __typename: 'StaticContentBlock',
        title: 'What are Atida cash rewards??',
        content: {
          json: {
            nodeType: 'document',
            data: {},
            content: [
              {
                nodeType: 'heading-2',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value: 'Nam venenatis',
                    marks: [],
                    data: {}
                  }
                ]
              },
              {
                nodeType: 'paragraph',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value:
                      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.',
                    marks: [],
                    data: {}
                  }
                ]
              }
            ]
          }
        },
        icon: {
          __typename: 'Icon',
          ref: 'iconReference'
        }
      },
      {
        __typename: 'StaticContentBlock',
        title: 'What are Atida cash rewards??',
        content: {
          json: {
            nodeType: 'document',
            data: {},
            content: [
              {
                nodeType: 'heading-2',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value: 'Nam venenatis',
                    marks: [],
                    data: {}
                  }
                ]
              },
              {
                nodeType: 'paragraph',
                data: {},
                content: [
                  {
                    nodeType: 'text',
                    value:
                      'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.',
                    marks: [],
                    data: {}
                  }
                ]
              }
            ]
          }
        },
        icon: {
          __typename: 'Icon',
          ref: 'iconReference'
        }
      }
    ]
  }
} as const

const graphQLMockSlider = {
  __typename: 'GroupOfStaticContentBlocks',
  title: 'Frequently asked questionsrs',
  type: 'Slider',
  staticContentBlocksCollection: {
    items: [contentBlockWithImage, contentBlockWithImage, contentBlockWithImage]
  }
} as const

const nullGraphQLMock = {
  __typename: 'GroupOfStaticContentBlocks',
  title: 'Frequently asked questions',
  contentBlocksCollection: {
    items: [null, null]
  }
} as const

describe(normalizeGroupOfStaticContentBlocks, () => {
  describe('GraphQL group of static content blocks', () => {
    it('pulls out the container of static content blocks - Accordion', () => {
      expect(
        normalizeGroupOfStaticContentBlocks(graphQLMockAccordion)
      ).toMatchSnapshot()
    })
    it('pulls out the container of static content blocks - Bullet points', () => {
      expect(
        normalizeGroupOfStaticContentBlocks(graphQLMockBulletPoints)
      ).toMatchSnapshot()
    })
    it('pulls out the container of content blocks with image - Slider', () => {
      expect(
        normalizeGroupOfStaticContentBlocks(graphQLMockSlider)
      ).toMatchSnapshot()
    })
  })
  describe('GraphQL null blocks', () => {
    it('does not error', () => {
      expect(
        normalizeGroupOfStaticContentBlocks(nullGraphQLMock)
      ).toMatchSnapshot()
    })
  })
})
