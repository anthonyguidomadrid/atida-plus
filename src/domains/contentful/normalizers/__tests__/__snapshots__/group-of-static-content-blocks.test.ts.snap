// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeGroupOfStaticContentBlocks GraphQL group of static content blocks pulls out the container of content blocks with image - Slider 1`] = `
Object {
  "blocks": Array [
    Object {
      "backgroundColor": "secondary-dark-sky-blue",
      "buttonPositionForTablet": "right",
      "buttonVariant": "primary",
      "buttonWidthForMobile": "auto",
      "content": "<p>We&#39;re here to help. Ask about our products or get expert medical advice.</p>",
      "contentType": "ContentBlockWithImage",
      "cta": Object {
        "content": "",
        "icon": "NavAdvice24",
        "label": "Chat with an expert",
        "url": "/contact",
      },
      "header": "",
      "image": Object {
        "description": "Person looking at a tablet",
        "title": "Contact image",
        "type": "image/jpeg",
        "url": "https://atida.com/adam-winger-iirbrh939yc-unsplash_1.jpg",
      },
      "imageOnTheLeft": false,
      "imageSize": "large",
      "isSponsoredContent": false,
      "showDescription": true,
      "textAlignmentForMobile": "left",
      "textColor": "primary-white",
      "textLink": Object {
        "content": "",
        "label": "Read our FAQ",
        "url": "/faq",
      },
      "title": "Any questions?",
      "titleTypography": "heading",
    },
    Object {
      "backgroundColor": "secondary-dark-sky-blue",
      "buttonPositionForTablet": "right",
      "buttonVariant": "primary",
      "buttonWidthForMobile": "auto",
      "content": "<p>We&#39;re here to help. Ask about our products or get expert medical advice.</p>",
      "contentType": "ContentBlockWithImage",
      "cta": Object {
        "content": "",
        "icon": "NavAdvice24",
        "label": "Chat with an expert",
        "url": "/contact",
      },
      "header": "",
      "image": Object {
        "description": "Person looking at a tablet",
        "title": "Contact image",
        "type": "image/jpeg",
        "url": "https://atida.com/adam-winger-iirbrh939yc-unsplash_1.jpg",
      },
      "imageOnTheLeft": false,
      "imageSize": "large",
      "isSponsoredContent": false,
      "showDescription": true,
      "textAlignmentForMobile": "left",
      "textColor": "primary-white",
      "textLink": Object {
        "content": "",
        "label": "Read our FAQ",
        "url": "/faq",
      },
      "title": "Any questions?",
      "titleTypography": "heading",
    },
    Object {
      "backgroundColor": "secondary-dark-sky-blue",
      "buttonPositionForTablet": "right",
      "buttonVariant": "primary",
      "buttonWidthForMobile": "auto",
      "content": "<p>We&#39;re here to help. Ask about our products or get expert medical advice.</p>",
      "contentType": "ContentBlockWithImage",
      "cta": Object {
        "content": "",
        "icon": "NavAdvice24",
        "label": "Chat with an expert",
        "url": "/contact",
      },
      "header": "",
      "image": Object {
        "description": "Person looking at a tablet",
        "title": "Contact image",
        "type": "image/jpeg",
        "url": "https://atida.com/adam-winger-iirbrh939yc-unsplash_1.jpg",
      },
      "imageOnTheLeft": false,
      "imageSize": "large",
      "isSponsoredContent": false,
      "showDescription": true,
      "textAlignmentForMobile": "left",
      "textColor": "primary-white",
      "textLink": Object {
        "content": "",
        "label": "Read our FAQ",
        "url": "/faq",
      },
      "title": "Any questions?",
      "titleTypography": "heading",
    },
  ],
  "contentType": "GroupOfStaticContentBlocks",
  "title": "Frequently asked questionsrs",
  "type": "Slider",
}
`;

exports[`normalizeGroupOfStaticContentBlocks GraphQL group of static content blocks pulls out the container of static content blocks - Accordion 1`] = `
Object {
  "blocks": Array [
    Object {
      "content": "<h2>Nam venenatis</h2><p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.</p>",
      "contentType": "StaticContentBlock",
      "title": "How do I earn Atida cash rewards?",
    },
    Object {
      "content": "<h2>Nam venenatis</h2><p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.</p>",
      "contentType": "StaticContentBlock",
      "title": "What are Atida cash rewards??",
    },
  ],
  "contentType": "GroupOfStaticContentBlocks",
  "title": "Frequently asked questionsrs",
  "type": "Accordion",
}
`;

exports[`normalizeGroupOfStaticContentBlocks GraphQL group of static content blocks pulls out the container of static content blocks - Bullet points 1`] = `
Object {
  "blocks": Array [
    Object {
      "content": "<h2>Nam venenatis</h2><p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.</p>",
      "contentType": "StaticContentBlock",
      "icon": "iconReference",
      "title": "How do I earn Atida cash rewards?",
    },
    Object {
      "content": "<h2>Nam venenatis</h2><p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.</p>",
      "contentType": "StaticContentBlock",
      "icon": "iconReference",
      "title": "What are Atida cash rewards??",
    },
    Object {
      "content": "<h2>Nam venenatis</h2><p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas et massa convallis, feugiat odio eget, mollis risus. Nunc a efficitur orci. Nam lacinia commodo ipsum, non blandit velit euismod in. Aliquam non dui nec tortor tincidunt consequat vel a massa. Suspendisse viverra luctus euismod. Donec vitae posuere leo. Ut posuere elementum libero eu fringilla. Suspendisse at pulvinar risus. Pellentesque eget bibendum dui, in feugiat odio. Maecenas fermentum ipsum sed tincidunt posuere.</p>",
      "contentType": "StaticContentBlock",
      "icon": "iconReference",
      "title": "What are Atida cash rewards??",
    },
  ],
  "contentType": "GroupOfStaticContentBlocks",
  "title": "Frequently asked questionsrs",
  "type": "Bullet points",
}
`;

exports[`normalizeGroupOfStaticContentBlocks GraphQL null blocks does not error 1`] = `
Object {
  "contentType": "GroupOfStaticContentBlocks",
  "title": "Frequently asked questions",
  "type": "",
}
`;
